robot/impl/local_robot_test.go

```diff
@@ -2426,116 +2426,224 @@ func TestDependentResources(t *testing.T) {
+					API:   gizmoAPI,
+				},
+			},
+			Services: []resource.Config{
+				{
+					Name:  "s",
+					Model: summationModel,
+					API:   summationAPI,
+				},
+			},
+		}
+		r.Reconfigure(ctx, cfg3)
 
-	// Remove module entirely.
-	cfg3 := &config.Config{
-		Components: []resource.Config{
-			{
-				Name:  "g",
-				Model: gizmoModel,
-				API:   gizmoAPI,
+		// Assert that adding module 'mod' back with original executable path re-adds
+		// modular component 'g' and modular service 's'.
+		r.Reconfigure(ctx, cfg)
+
+		_, err = r.ResourceByName(gizmoapi.Named("g"))
+		test.That(t, err, test.ShouldBeNil)
+		_, err = r.ResourceByName(summationapi.Named("s"))
+		test.That(t, err, test.ShouldBeNil)
+	})
+
+	t.Run("automatic reconfiguration", func(t *testing.T) {
+		cfg := &config.Config{
+			Modules: []config.Module{
+				{
+					Name:    "mod",
+					ExePath: rutils.ResolveFile("module/testmodule/testmodule"),
+				},
 			},
-		},
-		Services: []resource.Config{
-			{
-				Name:  "s",
-				Model: summationModel,
-				API:   summationAPI,
+			Components: []resource.Config{
+				{
+					Name:  "h",
+					Model: helperModel,
+					API:   generic.API,
+				},
 			},
-		},
-	}
-	r.Reconfigure(ctx, cfg3)
+		}
+		r.Reconfigure(ctx, cfg)
 
-	// Assert that adding module 'mod' back with original executable path re-adds
-	// modular component 'g' and modular service 's'.
-	r.Reconfigure(ctx, cfg)
+		h, err := r.ResourceByName(generic.Named("h"))
+		test.That(t, err, test.ShouldBeNil)
 
-	_, err = r.ResourceByName(gizmoapi.Named("g"))
-	test.That(t, err, test.ShouldBeNil)
-	_, err = r.ResourceByName(summationapi.Named("s"))
-	test.That(t, err, test.ShouldBeNil)
+		// Assert that removing testmodule binary and killing testmodule orphans
+		// helper 'h' a couple seconds after third restart attempt.
+		err = os.Remove(rutils.ResolveFile("module/testmodule/testmodule"))
+		test.That(t, err, test.ShouldBeNil)
+		_, err = h.DoCommand(ctx, map[string]interface{}{"command": "kill_module"})
+		test.That(t, err, test.ShouldNotBeNil)
+		test.That(t, err.Error(), test.ShouldContainSubstring,
+			"error reading from server")
+
+		// Check for "attempt 3" in logs every 100ms for 20s max.
+		waitForThirdAttempt := func() {
+			timer := time.NewTimer(20 * time.Second)
+			tick := time.NewTicker(100 * time.Millisecond)
+			defer timer.Stop()
+			defer tick.Stop()
+			for range tick.C {
+				select {
+				case <-timer.C:
+					t.Fatal("timed out waiting for 'attempt3' in server logs")
+				default:
+				}
+
+				if logs.FilterMessageSnippet("attempt 3").Len() > 0 {
+					time.Sleep(2 * time.Second)
+					break
+				}
+			}
+		}
+		waitForThirdAttempt()
+
+		_, err = r.ResourceByName(generic.Named("h"))
+		test.That(t, err, test.ShouldNotBeNil)
+		test.That(t, err, test.ShouldBeError,
+			resource.NewNotFoundError(generic.Named("h")))
+
+		/* Section below a WIP and awaiting input on desired behavior for
+		    * reconfiguration after module crash (unsuccessful restart)
```

Comment:
What do we expect users to do if their module crashes irrecoverably? I assumed they might see the error, remove the module from the config, and then re-add it all while keeping the server running, hoping that it now works? Is that something we need to support (i.e. resources should automatically re-associate with old module when they do that)? I can add something like the assertions below if so.

Reply:
Agreed. And to clarify, any time a user changes a config, (e.g. they fix a broken module) it should attempt to  (re)add any resources not fully ready. E.g. "h" could be broken for a while (perhaps due to a broken dependency OR a broken module) but when the problem is resolved (and if h is still in the config) there should be a try to load it.